### PR TITLE
[MPS] Fix linear backward SIGABRT with 1-D weight

### LINKED ISSUE #187988: [MPS] linear backward crashes (SIGABRT) with a 1-D weight: 'mps.matmul' op contracting dimensions differ
### 🐛 Describe the bug

`torch.nn.functional.linear` accepts a 1-D `weight` (shape `[in_features]`): it contracts and
drops the last input dimension, so the output loses that dimension. The forward pass works on
MPS, but the **backward** pass hard-crashes the process with a `SIGABRT` from MPSGraph graph
verification. The same computation is correct on CPU (and CUDA).

The crash cannot be caught — `PYTORCH_ENABLE_MPS_FALLBACK=1` does not help, because it is not a
missing op but an invalid lowering of an existing one.

#### Minimal reproduction

```python
import torch
import torch.nn.functional as F

x = torch.randn(3840, 33, device="mps", requires_grad=True)
w = torch.randn(33, device="mps", requires_grad=True)   # 1-D weight
y = F.linear(x, w)          # forward OK -> shape [3840]
y.sum().backward()          # <- SIGABRT in mps_linear_backward
```

A standard `nn.Linear` (2-D weight) is **not** affected; the crash is specific to a 1-D weight.

#### Error signature

```
(mpsFileLoc): .../MetalPerformanceShadersGraph/.../Core/Files/MPSGraphUtilities.mm:43:0:
error: 'mps.matmul' op contracting dimensions differ 3840 & 33
note: see current operation: %2 = "mps.matmul"(%arg1, %arg0)
  <{transpose_lhs = false, transpose_rhs = false}> : (tensor<3840xf32>, tensor<33xf32>) -> tensor<f32>
.../MPSGraphExecutable.mm:1484: failed assertion `original module failed verification'
```

Crashing thread (abbreviated):

```
at::native::mps_linear_backward(...)
  -> at::mps::MPSStream::executeMPSGraph(...)
  -> -[MPSGraphExecutable initWithGraph:...] -> commonPostInit -> MTLReportFailure -> abort (SIGABRT)
```

The two operand lengths are `N` (the flattened batch size of `grad_output`, which varies) and
`K` (the length of the 1-D weight). For `input [3840, 33]` they are `3840` and `33`.

#### Root cause

`mps_linear` (forward) explicitly supports a 1-D weight by unsqueezing it to `[1, K]`
(`weight_arg.dim() == 1 ? weight_arg.unsqueeze(0) : weight_arg`) and squeezing the trailing
output dimension back. `mps_linear_backward` does **not** mirror this: it feeds the raw 1-D
weight into `_mps_linear_backward_input`, where `matmul(grad_output, weight)` becomes a matmul of
two rank-1 tensors `(N,) x (K,)`. MPSGraph treats that as a vector dot-product (which requires
equal lengths) instead of the intended outer product, so graph verification aborts. The
`grad_weight` path is broken the same way.

A fix and a regression test (`test_mps.py::TestMPS::test_linear_1d_weight_backward`) are ready;
PR to follow.

### Versions

- PyTorch: 2.10.0 (also reproduces on current `main` — the relevant backward code is unchanged).
- Hardware: Apple M5, 24 GB unified memory (Mac17,4).
- OS: macOS 26.5.1 (25F80). Metal 373.2; MetalPerformanceShadersGraph 6.5.1.
- Python: 3.13.

cc @ezyang @gchanan @kadeng @msaroufim @albanD @gqchen @nikitaved @soulitzer @Varal7 @bobrenjc93 @malfet @aditvenk @snadampal @milpuz01 @aditew01 @nikhil-arm @fadara01 @robert-hardwick @nWEIdia @kulinseth @DenisVieriu97 @jhavukainen @Isalia20

### PR DESCRIPTION
**Pull request by adhi29**
Fixes #187988

## Problem

`F.linear` accepts a 1-D weight. The MPS forward handles it by lifting the weight to 2-D and squeezing the output (see `_mps_linear`), but `mps_linear_backward` fed the raw 1-D weight and 1-D grad_output straight into the 2-D matmul graphs. MPSGraph verification then failed and aborted the entire process with SIGABRT, which cannot be caught from Python:

```
error: 'mps.matmul' op contracting dimensions differ 3840 & 33
MPSGraphExecutable.mm:1484: failed assertion 'original module failed verification'
```

Reproduced on macOS with torch 2.13.0:

```python
x = torch.randn(3840, 33, device="mps", requires_grad=True)
w = torch.randn(33, device="mps", requires_grad=True)
F.linear(x, w).sum().backward()   # SIGABRT
```

## Fix

In `mps_linear_backward`, mirror what the forward does: unsqueeze the weight to `[1, in_features]` and grad_output to `[..., 1]`, run the regular 2-D backward graphs, and squeeze grad_weight back to 1-D.

While validating the fix on device I found that `_mps_linear_backward_input` also returns wrong values (zeros past the first 16KB page) for a non-contiguous grad_output when the reduction dimension is 1. This is reachable today without any 1-D weight:

```python
x = torch.randn(3840, 33, device="mps", requires_grad=True)
w = torch.randn(1, 33, device="mps")
F.linear(x, w).sum().backward()   # x.grad is mostly zeros, silently wrong
```

The unsqueezed grad_output from the 1-D path hits the same condition (the expanded stride-0 gradient that `sum()` produces), so the fix makes grad_output contiguous in `_mps_linear_backward_input`, the same guard the function already applies to the weight.

## Validation

The exact tensor operations the fixed code performs were validated on MPS against CPU reference gradients for shapes `(256, 33)`, `(3840, 33)` and `(4, 7, 33)`, including the stride-0 expanded gradient case. Max error observed was about 1e-5, well within the test tolerances.

Added `test_linear_1d_weight_backward` to `test/test_mps.py` next to the existing forward-only `test_linear_1d_weight`. The batch size 256 spans more than one 16KB page so a regression in the non-contiguous grad_output handling is also caught.

### CODE REVIEW

**Comment by adhi29** [comment]
Closing in favor of #187989, which covers this and was opened earlier. Missed it when I searched.
